master/internal/db/postgres_experiments.go

```diff
@@ -390,3 +394,146 @@ ORDER BY batches;`, &rows, metricName, trialID, startBatches, endBatches, startT
+		if err != nil {
+			return nil, errors.Wrap(err,
+				"failed to process training metrics for hyperparameter importance")
+		}
+		results = append(results, result)
+	}
+	return results, nil
+}
+
+// FetchHPImportanceValidationData retrieves all the data needed by the hyperparameter importance
+// algorithm to measure the relative importance of various hyperparameters for one specific
+// validation metric across all the trials in an experiment.
+func (db *PgDB) FetchHPImportanceValidationData(experimentID int, metric string) (
+	[]model.HPImportanceTrialData, error) {
+	var rows []hpImportanceDataWrapper
+	var results []model.HPImportanceTrialData
+	err := db.queryRows(`
+SELECT
+  t.id AS trial_id,
+  t.hparams AS hparams,
+  (s.prior_batches_processed + s.num_batches) AS batches,
+  v.metrics->'validation_metrics'->$1 as metric
+FROM trials t
+  INNER JOIN steps s ON t.id=s.trial_id
+  RIGHT JOIN validations v ON s.id=v.step_id AND s.trial_id=v.trial_id
+  INNER JOIN (
+    SELECT
+      t.id as trial_id,
+      s.id AS step_id,
+      max(s.prior_batches_processed + s.num_batches) AS batches
+    FROM trials t
+      INNER JOIN steps s ON t.id=s.trial_id
+      RIGHT JOIN validations v ON s.id=v.step_id AND s.trial_id=v.trial_id
+    WHERE t.experiment_id=$2
+      AND v.state = 'COMPLETED'
+    GROUP BY t.id, s.id
+  ) filter
+	ON s.id = filter.step_id
+	AND t.id = filter.trial_id`, &rows, metric, experimentID)
+	if err != nil {
+		return nil, errors.Wrapf(err,
+			"failed to get validation metrics for hyperparameter importance")
+	}
+	for _, row := range rows {
+		result, err := unmarshalHPImportanceHParams(row)
+		if err != nil {
+			return nil, errors.Wrap(err,
+				"Failed to process validation metrics for hyperparameter importance")
+		}
+		results = append(results, result)
+	}
+	return results, nil
+}
+
+// GetHPImportance returns the hyperparameter importance data and status for an experiment.
+func (db *PgDB) GetHPImportance(experimentID int) (result model.ExperimentHPImportance, err error) {
+	var jsonString []byte
+	err = db.sql.Get(&jsonString, "SELECT hpimportance FROM experiments WHERE id=$1", experimentID)
+	if err != nil {
+		return result, errors.Wrap(err, "Error retrieving hyperparameter importance")
+	}
+	if len(jsonString) > 0 {
+		err = json.Unmarshal(jsonString, &result)
+		if err != nil {
+			return result, errors.Wrap(err, "Error unmarshaling hyperparameter importance")
+		}
+	}
+	if result.TrainingMetrics == nil {
+		result.TrainingMetrics = make(map[string]model.MetricHPImportance)
+	}
+	if result.ValidationMetrics == nil {
+		result.ValidationMetrics = make(map[string]model.MetricHPImportance)
+	}
+	return result, err
+}
+
+// SetHPImportance writes the current hyperparameter importance data and status to the database.
+// It should only be called from the HPImportance manager actor, to ensure coherence.
+func (db *PgDB) SetHPImportance(experimentID int, value model.ExperimentHPImportance) error {
```

Comment:
I think if the structure really changes dramatically enough that we can't gracefully handle new / old in the code, we'd need to start using a different database column or something to distinguish between two "versions". But I'm pretty happy with the format being flexible and only depending on some pretty fundamental assumptions elsewhere in our product.

There's a JIRA to eventually add a scan on start up to restart any jobs that were in-progress when the master stopped. We could possibly piggy-back on that as a way to do more involved migrations if necessary.

Reply:
> I'm pretty happy with the format being flexible and only depending on some pretty fundamental assumptions 

Yeah, I definitely appreciate the simplicity.

> There's a JIRA to eventually add a scan on start up to restart any jobs that were in-progress when the master stopped. We could possibly piggy-back on that as a way to do more involved migrations if necessary.

Ok, that sounds like it could work great for that then. I'm just particularly cautious now about breaking changes, and this looks like an area we could easily make one by accident.